File: src/com/rian/difficultycalculator/beatmap/hitobject/sliderobject/SliderTick.java
package com.rian.difficultycalculator.beatmap.hitobject.sliderobject;

import com.rian.difficultycalculator.math.Vector2;

/**
 * Represents a slider tick.
 */
public class SliderTick extends SliderHitObject {
    /**
     * @param startTime              The time at which this slider tick starts, in milliseconds.
     * @param x                      The X position of the slider tick relative to the play field.
     * @param y                      The Y position of the slider tick relative to the play field.
     * @param spanIndex              The index of the span at which this slider hit object lies.
     * @param spanStartTime          The start time of the span at which this slider hit object lies, in milliseconds.
     */
    public SliderTick(double startTime, double x, double y, int spanIndex, double spanStartTime) {
        this(startTime, new Vector2(x, y), spanIndex, spanStartTime);
    }

    /**
     * @param startTime              The time at which this slider tick starts, in milliseconds.
     * @param position               The position of the slider tick relative to the play field.
     * @param spanIndex              The index of the span at which this slider hit object lies.
     * @param spanStartTime          The start time of the span at which this slider hit object lies, in milliseconds.
     */
    public SliderTick(double startTime, Vector2 position, int spanIndex, double spanStartTime) {
        super(startTime, position, spanIndex, spanStartTime);
    }

    /**
     * Copy constructor.
     *
     * @param source The source to copy from.
     */
    private SliderTick(SliderTick source) {
        this(source.startTime, source.position.x, source.position.y, source.spanIndex, source.spanStartTime);
    }

    @Override
    public SliderTick deepClone() {
        return new SliderTick(this);
    }
}

File: src/com/rian/osu/beatmap/hitobject/sliderobject/SliderTick.kt
package com.rian.osu.beatmap.hitobject.sliderobject

import com.rian.osu.GameMode
import com.rian.osu.beatmap.EmptyHitWindow
import com.rian.osu.beatmap.sections.BeatmapControlPoints
import com.rian.osu.beatmap.sections.BeatmapDifficulty
import com.rian.osu.math.Vector2
import kotlinx.coroutines.CoroutineScope

/**
 * Represents a slider tick.
 */
class SliderTick(
    /**
     * The time at which this [SliderTick] starts, in milliseconds.
     */
    startTime: Double,

    /**
     * The position of this [SliderTick] relative to the play field.
     */
    position: Vector2,

    /**
     * The index of the span at which this [SliderTick] lies.
     */
    private val spanIndex: Int,

    /**
     * The start time of the span at which this [SliderTick] lies, in milliseconds.
     */
    private val spanStartTime: Double
) : SliderHitObject(startTime, position) {
    override fun applyDefaults(
        controlPoints: BeatmapControlPoints,
        difficulty: BeatmapDifficulty,
        mode: GameMode,
        scope: CoroutineScope?
    ) {
        super.applyDefaults(controlPoints, difficulty, mode, scope)

        // Adding 200 to include the offset stable used.
        // This is so on repeats ticks don't appear too late to be visually processed by the player.
        val offset = if (spanIndex > 0) 200.0 else timePreempt * 0.66

        timePreempt = (startTime - spanStartTime) / 2 + offset
    }

    override fun createHitWindow(mode: GameMode) = EmptyHitWindow()
}
